### PR TITLE
Add write lock

### PR DESCRIPTION
**Pull request by casey**
I replaced the live write transaction `Mutex` with an `RwLock` so that writes can be initiated in parallel on different threads, but will block each other. I'm not sure if this is right, since there are a lot of subtleties with things like this.

Also, I was thinking that `WriteTransaction::new` should be safe, and should only take a reference to the database, and should grab the `RwLock` and allocate a new transaction ID, instead of spreading the safety concerns between `Database::begin_write` and `WriteTransaction::new`.

Depends on #370.

### CODE REVIEW

File: src/db.rs [hunk 1]
```diff
@@ -149,7 +150,7 @@ pub struct Database {
     mem: TransactionalMemory,
     next_transaction_id: AtomicTransactionId,
     live_read_transactions: Mutex<BTreeSet<TransactionId>>,
-    live_write_transaction: Mutex<Option<TransactionId>>,
+    live_write_transaction: RwLock<Option<TransactionId>>,
```
Comment (cberner) [inline src/db.rs:153]: I don't think this needs to change to an `RwLock`. You never call `.read()` and `.lock()` blocks just like `.write()`. Am I missing something?

File: src/db.rs [hunk 2]
```diff
@@ -434,13 +429,14 @@ impl Database {
     /// Returns a [`WriteTransaction`] which may be used to read/write to the database. Only a single
     /// write may be in progress at a time
     pub fn begin_write(&self) -> Result<WriteTransaction> {
```
Comment (cberner) [inline src/db.rs:431]: Can you update the docs to say that it will block if a write is in progress?

**Review by cberner — CHANGES_REQUESTED** [review]
This seems like a good change, but I had a few comments. Also, can you add a test that tests the blocking with two threads?

File: src/db.rs (hunk 1 above)
Comment (casey) [reply src/db.rs]: You're totally right. I think maybe I had a prior revision which required a RwLock for some reason.

File: src/db.rs (hunk 2 above)
Comment (casey) [reply src/db.rs:431]: Done.

**Comment by casey** [comment]
> This seems like a good change, but I had a few comments. Also, can you add a test that tests the blocking with two threads?

Comments addressed. Also, check out the "Make `begin_write` safe" commit. I think it's a slight improvement. At the expense of making two db fields `pub(crate)`, `WriteTransaction::new` is now safe, and the safety concerns are all in that function, instead of being spread across two functions.

**Comment by casey** [comment]
Except for the tests! Where should I put them?

**Comment by casey** [comment]
Added the test at the end of `basic_tests.rs`, let me know if there's a better place for it. This test consistently panics on master, but passes on the PR branch.

**Comment by cberner** [comment]
Yep, basic_tests or integration_tests is where I put most things. I should sort them better at some point :)

**Comment by cberner** [comment]
Looks like there's a compilation error. Also can you rebase to resolve the conflicts with your other PR?

**Comment by casey** [comment]
Rebased and resolved.

**Review by cberner — APPROVED** [review]

**Pull request merged**